src/uu/tail/src/chunks.rs

```diff
@@ -81,3 +92,424 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// use std::io::{BufReader, Cursor};
+    ///
+    /// let mut reader = BufReader::new(Cursor::new(""));
+    /// let num_print = 0;
+    /// let mut chunks = BytesChunkBuffer::new(num_print);
+    /// let mut iter = chunks.fill(&mut reader).unwrap();
+    ///
+    /// let chunk = iter.next();
+    /// assert!(chunk.is_none());
+    ///
+    /// let mut reader = BufReader::new(Cursor::new("a"));
+    /// let num_print = 1;
+    /// let mut chunks = BytesChunkBuffer::new(num_print);
+    /// let mut iter = chunks.fill(&mut reader).unwrap();
+    ///
+    /// let chunk = iter.next();
+    /// assert!(chunk.is_some());
+    /// assert_eq!(&[b'a'], chunk.unwrap().get_buffer());
+    /// assert_eq!(None, iter.next());
+    /// ```
+    pub(crate) fn fill(
+        &mut self,
+        reader: &mut BufReader<impl Read>,
+    ) -> UResult<Iter<Box<BytesChunk>>> {
+        let mut chunk = Box::new(BytesChunk::new());
+
+        // fill chunks with all bytes from reader and reuse already instantiated chunks if possible
+        while (chunk.fill(reader)?).is_some() {
+            self.bytes += chunk.bytes as u64;
+            self.chunks.push_back(chunk);
+
+            let first = &self.chunks[0];
+            if self.bytes - first.bytes as u64 > self.num_print {
+                chunk = self.chunks.pop_front().unwrap();
+                self.bytes -= chunk.bytes as u64;
+            } else {
+                chunk = Box::new(BytesChunk::new());
+            }
+        }
+
+        // quit early if there are no chunks for example in case the pipe was empty
+        if self.chunks.is_empty() {
+            return Ok(self.chunks.iter());
+        }
+
+        let chunk = self.chunks.pop_front().unwrap();
+        // calculate the offset in the first chunk and put the calculated chunk as first element in
+        // the self.chunks collection.
+        let offset = if self.num_print >= self.bytes {
+            // ignore a passed in value exceeding the number of actually read bytes and treat it
+            // like a value equal to the number of bytes.
+            0
+        } else {
+            // the calculated offset must be in the range 0 to BUFFER_SIZE and is therefore safely
+            // convertible to a usize without losses.
+            (self.bytes - self.num_print) as usize
+        };
+
+        self.chunks
+            .push_front(Box::new(BytesChunk::from(&chunk, offset)));
+
+        Ok(self.chunks.iter())
+    }
+}
+
+pub(crate) struct LinesChunk {
+    buffer: ChunkBuffer,
+    bytes: usize,
+    lines: usize,
+    delimiter: u8,
+}
+
+impl LinesChunk {
+    pub(crate) fn new(delimiter: u8) -> Self {
+        Self {
+            buffer: [0; BUFFER_SIZE],
+            bytes: 0,
+            lines: 0,
+            delimiter,
+        }
+    }
+
+    fn count_lines(&self) -> usize {
+        memchr::memchr_iter(self.delimiter, self.get_buffer()).count()
```

Comment:
What is the performance improvement of this over just `self.get_buffer().iter().filter(|c| c == seld.delimiter).count()`?

Reply:
Originally, I had the idea to search through the buffer with some fast c-like memory scan method. I know from the buffer, that it's a contiguous block and know the start and end location in the memory, so it looked appealing to me, but it seems to be unsafe in rust and I need more experience with rust to write unsafe code, I guess. I had no deeper look into memchr, but it was already in the secondary dependencies of uutils and after exchanging my slower for loop with it, I was happy and haven't asked any further. Maybe I missed something, however, I can try to exchange the dependency with your method, or you add it, if you like to.

Reply:
I checked the difference, now, and counting with the suggested filter method has the same performance penalties like my for loop had. It's only noticable when the number of lines passed to `-n` is close or even to `10_000_000` in the benchmark below with a 505MB file.

<details>
<summary>benchmark of <code>iter().filter(|c| c == seld.delimiter).count()</code></summary>

```
❯ hyperfine --warmup 3 --output pipe -L prg tail,target/release/tail -L values 10,1000,100000,10000000 -L option c,n 'cat random_ascii_505MB.BIG | {prg} -{option} +{values} -'
Benchmark 1: cat random_ascii_505MB.BIG | tail -c +10 -
  Time (mean ± σ):     305.9 ms ±   4.8 ms    [User: 71.6 ms, System: 417.4 ms]
  Range (min … max):   302.5 ms … 318.8 ms    10 runs

Benchmark 2: cat random_ascii_505MB.BIG | target/release/tail -c +10 -
  Time (mean ± σ):     310.2 ms ±   1.8 ms    [User: 70.5 ms, System: 423.0 ms]
  Range (min … max):   306.2 ms … 312.6 ms    10 runs

Benchmark 3: cat random_ascii_505MB.BIG | tail -c +1000 -
  Time (mean ± σ):     310.2 ms ±  13.2 ms    [User: 66.7 ms, System: 421.8 ms]
  Range (min … max):   301.1 ms … 346.8 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 4: cat random_ascii_505MB.BIG | target/release/tail -c +1000 -
  Time (mean ± σ):     309.9 ms ±   2.5 ms    [User: 51.5 ms, System: 441.0 ms]
  Range (min … max):   306.3 ms … 314.9 ms    10 runs

Benchmark 5: cat random_ascii_505MB.BIG | tail -c +100000 -
  Time (mean ± σ):     305.1 ms ±   2.1 ms    [User: 67.8 ms, System: 420.4 ms]
  Range (min … max):   302.8 ms … 309.3 ms    10 runs

Benchmark 6: cat random_ascii_505MB.BIG | target/release/tail -c +100000 -
  Time (mean ± σ):     309.1 ms ±   2.2 ms    [User: 56.0 ms, System: 436.2 ms]
  Range (min … max):   305.8 ms … 312.5 ms    10 runs

Benchmark 7: cat random_ascii_505MB.BIG | tail -c +10000000 -
  Time (mean ± σ):     302.9 ms ±   1.5 ms    [User: 69.9 ms, System: 414.8 ms]
  Range (min … max):   299.7 ms … 305.8 ms    10 runs

Benchmark 8: cat random_ascii_505MB.BIG | target/release/tail -c +10000000 -
  Time (mean ± σ):     307.2 ms ±   1.7 ms    [User: 61.2 ms, System: 428.1 ms]
  Range (min … max):   304.6 ms … 310.9 ms    10 runs

Benchmark 9: cat random_ascii_505MB.BIG | tail -n +10 -
  Time (mean ± σ):     305.0 ms ±   2.7 ms    [User: 60.8 ms, System: 427.3 ms]
  Range (min … max):   301.5 ms … 309.5 ms    10 runs

Benchmark 10: cat random_ascii_505MB.BIG | target/release/tail -n +10 -
  Time (mean ± σ):     314.8 ms ±  10.4 ms    [User: 66.2 ms, System: 427.1 ms]
  Range (min … max):   305.9 ms … 339.7 ms    10 runs

Benchmark 11: cat random_ascii_505MB.BIG | tail -n +1000 -
  Time (mean ± σ):     306.2 ms ±   2.4 ms    [User: 73.9 ms, System: 415.8 ms]
  Range (min … max):   303.3 ms … 310.3 ms    10 runs

Benchmark 12: cat random_ascii_505MB.BIG | target/release/tail -n +1000 -
  Time (mean ± σ):     309.9 ms ±   2.9 ms    [User: 71.6 ms, System: 422.2 ms]
  Range (min … max):   308.2 ms … 318.0 ms    10 runs

Benchmark 13: cat random_ascii_505MB.BIG | tail -n +100000 -
  Time (mean ± σ):     305.3 ms ±   3.3 ms    [User: 76.0 ms, System: 412.5 ms]
  Range (min … max):   302.0 ms … 311.6 ms    10 runs

Benchmark 14: cat random_ascii_505MB.BIG | target/release/tail -n +100000 -
  Time (mean ± σ):     314.4 ms ±  15.0 ms    [User: 73.2 ms, System: 420.3 ms]
  Range (min … max):   307.1 ms … 356.7 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 15: cat random_ascii_505MB.BIG | tail -n +10000000 -
  Time (mean ± σ):     205.0 ms ±   2.6 ms    [User: 61.2 ms, System: 291.8 ms]
  Range (min … max):   199.7 ms … 207.4 ms    14 runs

Benchmark 16: cat random_ascii_505MB.BIG | target/release/tail -n +10000000 -
  Time (mean ± σ):     337.6 ms ±   7.8 ms    [User: 223.7 ms, System: 300.3 ms]
  Range (min … max):   329.7 ms … 353.1 ms    10 runs

Summary
  'cat random_ascii_505MB.BIG | tail -n +10000000 -' ran
    1.48 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -c +10000000 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -n +10 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -c +100000 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -n +100000 -'
    1.49 ± 0.03 times faster than 'cat random_ascii_505MB.BIG | tail -c +10 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -n +1000 -'
    1.50 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +10000000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +100000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +1000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +1000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +10 -'
    1.51 ± 0.07 times faster than 'cat random_ascii_505MB.BIG | tail -c +1000 -'
    1.53 ± 0.08 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +100000 -'
    1.54 ± 0.05 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +10 -'
    1.65 ± 0.04 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +10000000 -'

```

</details>

Reply:
Alright, thank you for investigating! `memchr` sounds like a good option then.